cEP-0015.md

```diff
@@ -0,0 +1,98 @@
+
+## Abstract
+
+This cEP describes the details of enhancement of
+[corobo](https://github.com/coala/corobo) in terms of security, tests,
+configurability and the new plugins that are to be added to corobo as part
+of the
+[GSoC Project](https://summerofcode.withgoogle.com/projects/#6603667076022272).
+
+## Security
+
+### Why improve security?
+
+Security has been one of the major concerns due to some past experiences.
+We want to provide access controls based on user’s GitHub team memberships.
+
+### How to improve security?
+
+1. Make all LabHub commands except invite require being a member of the
+   organization.
+
+2. Add ability to ban from all gitter rooms at once.
+
+3. Remove auto-invite and give invite permissions to developers.
+
+4. Prevent newcomers from self-assigning themselves and give developers
+   permission to assign newcomers.
+
+5. Prevent corobo being used to spam a room.
+
+6. Force newcomers to finish one issue first.
+
+7. Require newcomers to find a newcomer issue to work on before they are
+   invited to join the organization.
+
+## Tests
+
+### Why improve existing tests and infrastructure?
+
+The default TestBase provided by errbot is not enough for testing plugins like
+LabHub, which required intensive mocking. One of the examples of this case is
+issues like [this](https://github.com/errbotio/errbot/issues/1086).
+
+### How to improve existing tests and infrastructure?
+
+This will involve making changes upstream in
+[Errbot](https://github.com/errbotio/errbot/) and extend the existing testing
+infrastructure to implement better testing for plugins like LabHub.
+
+## Configurability
+
+### Why make corobo more configurable?
+
+corobo has a potential to be used by other organizations for similar tasks like
+onboarding and automation. Currently, it is not configurable and many plugins
+are still very coala specific. Making it more configurable will allow other
+organizations to adapt corobo to cater their needs.
+
+### How to make corobo more configurable?
+
+By making existing coala specific plugins generic for other organizations we
+can ensure configurability.
+
+List of Plugins which are coala specific and can be generalized:
+
+1. `LabHub (invite, assign, unassign, mark wip|pending, pr_stats)` plugin are
+   meant to work for a specific team names like maintainers, developers,
+   newcomers and coala repositories.
+
+2. `explain` currently uses hardcoded coala explainations.
+
+3. `coatils (contrib_stats)` shows stats of contributions made to coala.
+
+4. `searchdocs` uses API_DOCS and USER_DOCS links which are constant for coala.
+
+## New Features
+
+### Suggest newcomer issues
+
+```
+> corobo suggest issues newcomer
+  Here are some newcomer issues:
+
+  coala/coala#5276 [Modify installing from git instructions]
+  coala/coala-bears#2376 [Fix invalid language setting]
+  coala/coala-quickstart#128 [add .DS_Store to .gitignore]
+  coala/coala-bears#2326 [Add .pytest_cache to .gitignore]
+```
```

Comment:
would be great if you can stuff some more implementation details into this 😉

Reply:
No there is no issue yet, I'll create one and mention it here.

Reply:
not about this feature, kind of comparing this cEP in general to others

Reply:
https://github.com/coala/corobo/issues/528 has been created.

IMO this cEP, and probably even this project, should not have new features which are not related to "security, tests and configurability".  Your project summary in GSoC doesnt mention new features.  Your GSoC proposal did mention adding this feature, but it definitely isnt relevant to the stated scope of this cEP.

It might be still included in your milestone tasks, but it may be dropped if the project team agrees that other pending work is more important. e.g. one of the critical components of this project takes more time than expected. This feature is useless if the corobo hardening isnt completed as best as is possible.

Anyone could add new features like this.  It doesnt require a GSoC student to do it.  GCI students last year built bigger more complicated plugins in less than three days.  If you dont build it this GSoC, no drama .. a newcomer or GCI student will build it by the end of the year.

Also I feel that it would be inappropriate to build more GitHub-only features into corobo.  corobo should exclusively use IGitt, and all plugins should work in all hosters supported by IGitt, unless it is functionality which is only supported by one hoster and it is really critical that it is available in corobo.

fwiw, another [2018 GSoC project](https://github.com/coala/projects/commit/f5ca3b597f9693fd82e7e71072deda63abed701a) will be providing a GitHub+GitLab issue search tool.

Reply:
Okay, I'll remove this section since it is out of scope for this project :)